Feature/user management + updated auth

markbotterill opened the pull request:
Reviewer: @sidravi1 
Estimate: 1 hours

---

## Ticket

Fixes: [JIRA_TICKET_LINK](https://lastmilehealth.atlassian.net/browse/AFHC-7)

## Description

Add users table, add user login endpoint, add two methods of verification (keep API_KEY but also enable user login + JWT usage).

Add tests

### Changes

Added users router. 
Modified `dependencies.py`
Ingestion endpoints now have "admin_only" lock on them.

### Future Tasks (optional)

Discuss integration w/ Google OAuth if needed
Discuss which endpoints need `admin_only`

## How has this been tested?

New tests written, all pass.

Can create a new user in Swagger UI, using API_KEY, then can log out on Swagger and log back in with email + password combination instead. 

Can test adding + deleting users.

## Checklist

Fill with `x` for completed.

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts

(Delete any items below that are not relevant)
- [x] I have updated the automated tests
- [ ] I have updated the scripts in `scripts/`
- [x] I have updated the requirements
- [ ] I have updated the README file
- [ ] I have updated affected documentation
- [ ] I have updated the CI/CD scripts in `.github/workflows/`
- [ ] I have updated the Terraform code

markbotterill committed:
Add users table migration

markbotterill committed:
Add auth routers + utils

markbotterill committed:
Cleanup

markbotterill committed:
Added jose for JWT to requirements

markbotterill committed:
Add delete functionality + updated user info schema

markbotterill committed:
Added stubs for jose

markbotterill committed:
Change import path

markbotterill committed:
Add passlib stubs

markbotterill committed:
Modify schemas to avoid mypy error

markbotterill committed:
Add status code to HTTPException

markbotterill committed:
Remove await on add

markbotterill committed:
Add emailvalidator to requirements

markbotterill committed:
Add explicit type to satisfy mypy

markbotterill committed:
Add passlib to reqs

markbotterill committed:
Simplify dict usage

markbotterill committed:
Include lambda function to create callable instead

markbotterill committed:
Repeat the same for chat schema

markbotterill committed:
Remove passlib only use bcrypt

markbotterill committed:
Remove passlib only use bcrypt in utils

markbotterill committed:
Add more graceful error handling for existing user

markbotterill committed:
Add tests for users

markbotterill committed:
Add bcrypt to module ignore in pyproject.toml

markbotterill committed:
Add bcrypt to requirements

markbotterill committed:
Merge branch 'develop' into feature/user-management

sidravi1 committed:
Add multi-turn chat endpoint (#33)

* Chat: single turn only (#19)
* Add history to chat to allow multi-turn conversation (#22)
* [After #22] Chat: Added Cross encoder (#23)
* Chat: Fixes from tech session (#28)
* Chat docs (#27)

sidravi1 committed:
fixed lifespan (#35)

sidravi1 commented on backend/app/auth/config.py:
```diff
@@ -1,3 +1,7 @@
 
 API_SECRET_KEY = os.getenv("API_SECRET_KEY", "my_secret_key")
+# Authentication and Authorization
+SECRET_KEY = os.environ.get("SECRET_KEY", "dummy-secret-key")
```
Can we please call it `JWT_SECRET_KEY`? This is an important one orgs shouldn't forget to update to let's also add it to `template.backend.env`

sidravi1 commented on backend/app/auth/dependencies.py:
```diff
@@ -1,26 +1,74 @@
     HTTPBearer,
+    OAuth2PasswordBearer,
 )
+from jose import JWTError, jwt
```
`jose` is no longer maintained. Can i suggest switching to [pyjwt](https://github.com/jpadilla/pyjwt)?

sidravi1 replied:
Actually, worth checking out the AAQ repo for auth implementation

sidravi1 commented on backend/app/auth/config.py:
```diff
@@ -1,3 +1,7 @@
 API_SECRET_KEY = os.getenv("API_SECRET_KEY", "my_secret_key")
+# Authentication and Authorization
+SECRET_KEY = os.environ.get("SECRET_KEY", "dummy-secret-key")
+ALGORITHM = os.environ.get("ALGORITHM", "HS256")
```
And call this `JWT_ALGORITHM`?

sidravi1 commented on backend/app/auth/dependencies.py:
```diff
@@ -1,26 +1,74 @@
+    token: str = Security(oauth2_scheme),
+    credentials: HTTPAuthorizationCredentials = Security(api_key_scheme),
+    session: AsyncSession = Depends(get_async_session),
+) -> Optional[Users]:
```
You can now also describe optional type as `Users | None`

sidravi1 commented on backend/app/auth/dependencies.py:
```diff
@@ -1,26 +1,74 @@
+                if user:
+                    return user
+        except JWTError:
+            pass
```
So here, the token was passed but it was incorrect. We should log this as a warning that includes whatever was in exception message

sidravi1 commented on backend/app/auth/dependencies.py:
```diff
@@ -1,26 +1,74 @@
+    """
+    This dependency function authenticates users using either JWT or API key."""
+    user: Optional[Users] = None
+    # Try API Key authentication
```
Why do we keep the API Key auth? We can drop it right?
I guess an argument to keep it is if we ever connect a front end that can't login. But I'd put it back in when that is needed.

sidravi1 commented on backend/app/users/models.py:
```diff
@@ -0,0 +1,20 @@
+
+    user_id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
+    email: Mapped[str] = mapped_column(String, unique=True, nullable=False)
+    password: Mapped[str] = mapped_column(String, nullable=False)
```
Can we call this `hashed_password`?

sidravi1 commented on backend/app/users/utils.py:
```diff
@@ -0,0 +1,19 @@
+import bcrypt
```
Cool! Didn't know about this library. We implemented this from scratch on AAQ.

sidravi1 commented on backend/app/auth/schemas.py:
```diff
@@ -0,0 +1,20 @@
+    Class representing the token data model.
+    """
+
+    user_id: Optional[str] = None
```
Why optional? Do we sometimes send back `TokenData` with an empty `user_id`?

sidravi1 commented on backend/app/auth/utils.py:
```diff
@@ -0,0 +1,75 @@
+async def get_current_user(
+    token: str = Depends(oauth2_scheme),
+    session: AsyncSession = Depends(get_async_session),
+    role_check: Optional[str] = None,
```
How about using enums for roles? say `ADMIN` and `USER`?

sidravi1 commented on backend/app/users/models.py:
```diff
@@ -0,0 +1,20 @@
+    created_at: Mapped[datetime] = mapped_column(
+        TIMESTAMP(timezone=True), server_default=text("now()"), nullable=False
+    )
+    role: Mapped[str] = mapped_column(String, nullable=False, default="user")
```
You can use the enum types here as well 

https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map

sidravi1 commented on backend/app/auth/utils.py:
```diff
@@ -0,0 +1,75 @@
+    return token_data
+
+
+async def get_current_user(
```
What calls this function? if not used, can we remove it?

sidravi1 replied:
Also, it has the same name as the one in `auth.dependencies`. Can we change the name if we are keeping it?

sidravi1 commented on backend/app/chat/schemas.py:
```diff
@@ -48,7 +48,7 @@ class ChatResponse(BaseModel):
     )
     response_metadata: Optional[dict] = Field(
-        default_factory=dict, examples=[{"timestamp": "2024-10-16T12:31:00Z"}]
+        default_factory=lambda: {}, examples=[{"timestamp": "2024-10-16T12:31:00Z"}]
```
Hmmm... i thought I fixed this in my PR.

markbotterill replied:
Yeah I can't find it in codebase any more - should go away when I update to `develop`

markbotterill replied:
Sorted with a rebase 🫡

sidravi1 commented on backend/app/chat/routers.py:
```diff
@@ -17,7 +17,7 @@
 )
 
-router = APIRouter(dependencies=[Depends(authenticate_key)], tags=["Chat endpoints"])
+router = APIRouter(dependencies=[Depends(get_current_user)], tags=["Chat endpoints"])
```
We can now also get userid from the token and not require it in the `/chat` request schema

sidravi1 replied:
ideally, we also have a test case where we access the endpoints with username/password auth. And with both user and admin roles.

markbotterill replied:
@sidravi1 Can we implement the `user_id` switcherooo in a later PR? I fear this one is getting quite large.

sidravi1 commented on backend/app/users/routers.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    Create a new user.
+    """
+    if user_data.role not in ["admin", "user", None]:
```
Let's use Enum!

sidravi1 replied:
And add it to schema for UserCreate so it can do this check for us

sidravi1 commented on backend/app/users/routers.py:
```diff
@@ -0,0 +1,87 @@
+    session.add(new_user)
+    try:
+        await session.commit()
+    except IntegrityError:
```
can we get the IntegrityError as `e` and then raise HTTPException from `e`? More detail is good!

sidravi1 commented on backend/app/users/schemas.py:
```diff
@@ -0,0 +1,43 @@
+    )
+
+
+class UserOut(BaseModel):
```
What about `NewUser` as the name of this class? `UserOut` doesn't seem informative

sidravi1 replied:
Or maybe just `User`?

markbotterill replied:
So the logic here - and I was basically following directly from a tutorial but it made sense to me - was that any time we output users from the DB we should never have their password shown. So basically any `get`/`post` operations we do should follow this schema. So `UserOut` represents showing a User's info publicly.

I do use `UserOut` for other calls -> on creation and deletion too. It basically serves as a nice template for "Hey I did a thing with this user (no password shown) and this was what it was"

sidravi1 commented on backend/app/users/models.py:
```diff
@@ -0,0 +1,20 @@
+from ..models import Base
+
+
+class Users(Base):
```
The convention we have been following is to add `DB` at the end so its not confused with the schema. So this should be `UsersDB`

sidravi1 commented on backend/app/users/routers.py:
```diff
@@ -0,0 +1,87 @@
+}
+
+router = APIRouter(
+    dependencies=[Depends(authenticate_either)],
```
Should only admins we allowed to create users?

markbotterill replied:
I think - for now - yes? Seems logical to me

sidravi1 commented on backend/app/users/models.py:
```diff
@@ -0,0 +1,20 @@
+
+    __tablename__ = "users"
+
+    user_id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
```
Maybe for a future PR, we should keep track of when the user logged in and logged out. Probably in a separate table.

markbotterill replied:
Agreed for a future PR! I will add to tech session agenda.

sidravi1 commented on backend/migrations/versions/2024_11_21_d40e086dadb5_add_users_table.py:
```diff
@@ -0,0 +1,46 @@
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+        sa.Column("role", sa.String(), nullable=True),
```
model has this as `nullable=False`. Should be the same here right?

sidravi1 replied:
And if you do end up using Enum, you'll have to update it here as well

sidravi1 commented on backend/migrations/versions/2024_11_21_d40e086dadb5_add_users_table.py:
```diff
@@ -0,0 +1,46 @@
+        sa.PrimaryKeyConstraint("user_id"),
+        sa.UniqueConstraint("email"),
+    )
+    op.create_index(op.f("ix_users_user_id"), "users", ["user_id"], unique=False)
```
Should this be `unique=True`?

sidravi1 commented on backend/requirements.txt:
```diff
@@ -8,3 +8,7 @@ PyPDF2==3.0.1
 sentence-transformers==3.2.0
 python-multipart==0.0.12
+python-jose==3.3.0
+types-python-jose==3.3.4.20240106
```
Types should go in `requirements-dev.txt`

sidravi1 commented on backend/requirements.txt:
```diff
@@ -8,3 +8,7 @@ PyPDF2==3.0.1
 pgvector==0.3.5
 sentence-transformers==3.2.0
 python-multipart==0.0.12
+python-jose==3.3.0
```
See note about switching to `pyjwt`

sidravi1 commented on backend/app/users/routers.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    Delete a user by user_id.
+    """
+    result = await session.execute(select(Users).where(Users.user_id == user_id))
```
You were going to make this a soft delete right? That's probably good idea so we avoid orphan records etc.

sidravi1 commented on backend/tests/api/test_users.py:
```diff
@@ -0,0 +1,132 @@
+from app.users.routers import router
+from fastapi.testclient import TestClient
+
+client = TestClient(router)
```
You can get this from `conftest.py`.

sidravi1 commented on backend/tests/api/test_users.py:
```diff
@@ -0,0 +1,132 @@
+client = TestClient(router)
+
+
+def test_create_user(client: TestClient) -> None:
```
The biggest issue is that the tests dont clean up after themselves. So they may start interfering with other tests as content will remain the db.

Suggest using a fixture that just yields and the cleans the db by running a DELETE SQL statement

sidravi1 commented on backend/tests/api/test_users.py:
```diff
@@ -0,0 +1,132 @@
+
+
+def test_get_user(client: TestClient) -> None:
+    headers = {
```
Suggest using a fixture like I have in `test_chat.py`

sidravi1 commented on backend/tests/api/test_users.py:
```diff
@@ -0,0 +1,132 @@
+        "password": "securepassword",
+        "role": "user",
+    }
+    create_response = client.post("/users/create", json=user_data, headers=headers)
```
Should have a non-admin try to create this and fail

markbotterill replied:
I think - for now - yes?

sidravi1 reviewed (changes requested):
Thanks @markbotterill! This is big lift. Very nice work :)

Left a few comments to resolve. Also, need to rebase from develop and add soft delete of users. Happy to review again tomorrow evening.

sidravi1 reviewed (approved):
Thanks @robela!

Docker containers build and deploy! Look forward to the integration.
You'll need to rebase from `develop`

markbotterill committed:
SECRET_KEY -> JWT_SECRET_KEY globally

markbotterill committed:
Remove stubs + jose -> PyJWT

markbotterill committed:
Improve error logging for JWT + switch to PyJWT

markbotterill committed:
password -> hashed_password

robela committed:
AFHC 58: Set up a Docker container for the Chainlit (#34)

* servies layer Structure
Modularity: Each service handles a specific functionality, making the app easier to maintain.

* remove lib

* mypy fix

* mypy fix

* comment added

* comment-added

* comment addded

* comment

* comment

* fix and up and running

---------

Co-authored-by: LMH-L-7092\ralazar <ralazar@lastmilehealth.org>

markbotterill committed:
Switch to enum

markbotterill committed:
Users -> UsersDB + soft delete

markbotterill committed:
Undo accidental use of DB pw in URL create call

markbotterill committed:
Add enum to migration script

markbotterill committed:
Use enum in SQLAlchemy model, too

markbotterill committed:
Switch to enum in schemas

markbotterill committed:
Modifying alembic script

markbotterill committed:
Correct alembic migration script

markbotterill committed:
Add await

markbotterill committed:
Refine enum definition

markbotterill committed:
First admin script draft

robellmh committed:
fix for chillit

robellmh committed:
Merge branch 'feature/AFHC-61_API-Integration' of https://github.com/robellmh/HCD-AI into feature/AFHC-61_API-Integration

markbotterill committed:
pull updates

markbotterill committed:
Make user_id index unique

markbotterill committed:
Match backend scripts + models to Alembic migration implementation

markbotterill committed:
Changed language in deletion protocol

markbotterill committed:
Refine IntegrityError handling

markbotterill committed:
Removed unused app utils

markbotterill committed:
Remove await on add

markbotterill committed:
add action taken field to schema for UserOut

robellmh committed:
chilit update logo and icon

robellmh committed:
Merge branch 'develop' of https://github.com/robellmh/HCD-AI into feature/AFHC-61_API-Integration

robellmh committed:
setting changed

robela committed:
Merge pull request #1 from Last-Mile-Health-DH/feature/AFHC-61_API-Integration

Feature/afhc 61 API integration

markbotterill committed:
Merge remote-tracking branch 'new-repo/develop' into feature/user-management

markbotterill committed:
Add .chainlit folder to git ignore

markbotterill committed:
Add API_URL + refactor error logging + add chat_id

markbotterill committed:
Create config file at top level

markbotterill committed:
First draft of app.py - add JWT login usage + chat_id history

markbotterill committed:
Replace os.getenv w/ ..config imports

markbotterill committed:
Add dummy str(0) and fix types

markbotterill committed:
Remove user_id from chat endpoint - get from token

markbotterill committed:
Drop user_id from frontend call

markbotterill committed:
user_id as int NOT as string

markbotterill committed:
Fix relative import

markbotterill committed:
Add template .env for chainlit app

markbotterill committed:
Mod docstring since user_id no longer used in call

markbotterill committed:
Needless metadata included in frontend
